src/main/java/com/google/api/codegen/config/DiscoveryField.java

```diff
@@ -23,36 +23,93 @@
 import com.google.api.codegen.discovery.Schema.Type;
 import com.google.api.codegen.transformer.ImportTypeTable;
 import com.google.api.codegen.util.Name;
+import com.google.api.codegen.util.SymbolTable;
 import com.google.api.codegen.util.TypeName;
 import com.google.api.tools.framework.model.Oneof;
 import com.google.api.tools.framework.model.TypeRef.Cardinality;
 import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
 import com.google.common.collect.ImmutableList;
+import java.util.AbstractMap;
+import java.util.Comparator;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import javax.annotation.Nullable;
 
 /** A field declaration wrapper around a Discovery Schema. */
 public class DiscoveryField implements FieldModel, TypeModel {
   private final List<DiscoveryField> properties;
+  private final DiscoApiModel apiModel;
   // Dereferenced schema to use for rendering type names and determining properties, type, and format.
   private final Schema schema;
-
   // Not dereferenced schema; used in rendering this FieldModel's parameter name.
   private final Schema originalSchema;
-  private final DiscoApiModel apiModel;
+
+  // Unformatted type name for this Field.
+  // For message-type Fields, this will be unique per namespace, as defined by apiModel.
+  private final String typeName;
+
+  // Comparator for Schemas that have children schemas.
+  private static Comparator<Schema> messageSchemaComparator =
+      (Schema s1, Schema s2) -> s1.getIdentifier().compareTo(s2.getIdentifier());
+
+  // Comparator for DiscoveryFields.
+  private static Comparator<Map.Entry<Schema, DiscoApiModel>> discoveryFieldComparator =
+      new Comparator<Map.Entry<Schema, DiscoApiModel>>() {
+        @Override
+        public int compare(
+            Map.Entry<Schema, DiscoApiModel> o1, Map.Entry<Schema, DiscoApiModel> o2) {
+          if (o1.getValue().getDefaultPackageName().equals(o2.getValue().getDefaultPackageName())) {
+            // if the namespaces are the same, use the schema's hashcode to differentiate.
+            return o1.getKey().hashCode() - o2.getKey().hashCode();
+          }
+          return o1.getValue()
+              .getDefaultPackageName()
+              .compareTo(o2.getValue().getDefaultPackageName());
+        }
+      };
+  private static Map<Map.Entry<Schema, DiscoApiModel>, DiscoveryField> globalObjects =
```

Comment:
This is... quite unconventional (use a map entry as a key). I would suggest making your own data class for the key.